src/poll.rs

```diff
@@ -179,6 +179,180 @@ libc_bitflags! {
+    ///
+    /// > Specifying  a timeout of zero causes poll() to return immediately, even if no file
+    /// > descriptors are ready.
+    pub const ZERO: Self = Self(0);
+    /// Blocks for at most [`std::i32::MAX`] milliseconds.
+    pub const MAX: Self = Self(i32::MAX);
+    /// Returns if `self` equals [`PollTimeout::NONE`].
+    pub fn is_none(&self) -> bool {
+        // > Specifying a negative value in timeout means an infinite timeout.
+        *self <= Self::NONE
+    }
+    /// Returns if `self` does not equal [`PollTimeout::NONE`].
+    pub fn is_some(&self) -> bool {
+        !self.is_none()
+    }
+    /// Returns the timeout in milliseconds if there is some, otherwise returns `None`.
+    pub fn timeout(&self) -> Option<i32> {
+        self.is_some().then_some(self.0)
+    }
+}
+
+impl<T: Into<PollTimeout>> From<Option<T>> for PollTimeout {
+    fn from(x: Option<T>) -> Self {
+        x.map_or(Self::NONE, |x| x.into())
+    }
+}
+impl TryFrom<Duration> for PollTimeout {
+    type Error = <i32 as TryFrom<u128>>::Error;
+    fn try_from(x: Duration) -> std::result::Result<Self, Self::Error> {
+        Ok(Self(i32::try_from(x.as_millis())?))
+    }
+}
+impl TryFrom<u128> for PollTimeout {
+    type Error = <i32 as TryFrom<u128>>::Error;
+    fn try_from(x: u128) -> std::result::Result<Self, Self::Error> {
+        Ok(Self(i32::try_from(x)?))
+    }
+}
+impl TryFrom<u64> for PollTimeout {
+    type Error = <i32 as TryFrom<u64>>::Error;
+    fn try_from(x: u64) -> std::result::Result<Self, Self::Error> {
+        Ok(Self(i32::try_from(x)?))
+    }
+}
+impl TryFrom<u32> for PollTimeout {
+    type Error = <i32 as TryFrom<u32>>::Error;
+    fn try_from(x: u32) -> std::result::Result<Self, Self::Error> {
+        Ok(Self(i32::try_from(x)?))
+    }
+}
+impl From<u16> for PollTimeout {
+    fn from(x: u16) -> Self {
+        Self(i32::from(x))
+    }
+}
+impl From<u8> for PollTimeout {
+    fn from(x: u8) -> Self {
+        Self(i32::from(x))
+    }
+}
+impl TryFrom<i128> for PollTimeout {
+    type Error = <i32 as TryFrom<i128>>::Error;
+    fn try_from(x: i128) -> std::result::Result<Self, Self::Error> {
+        match x {
+            // > Specifying a negative value in timeout means an infinite timeout.
+            i128::MIN..=-1 => Ok(Self::NONE),
+            millis @ 0.. => Ok(Self(i32::try_from(millis)?)),
+        }
+    }
+}
+impl TryFrom<i64> for PollTimeout {
+    type Error = <i32 as TryFrom<i64>>::Error;
+    fn try_from(x: i64) -> std::result::Result<Self, Self::Error> {
+        match x {
+            i64::MIN..=-1 => Ok(Self::NONE),
+            millis @ 0.. => Ok(Self(i32::try_from(millis)?)),
+        }
+    }
+}
+impl From<i32> for PollTimeout {
+    fn from(x: i32) -> Self {
+        Self(x)
+    }
+}
+impl From<i16> for PollTimeout {
+    fn from(x: i16) -> Self {
+        Self(i32::from(x))
+    }
+}
+impl From<i8> for PollTimeout {
+    fn from(x: i8) -> Self {
+        Self(i32::from(x))
+    }
+}
```

Comment:
The conversions from signed integers seem to be insistent, for `i128/i64`, a negative number would be converted to `PollTimeOut(-1)`, and for `i32/i16/i8`, they are stored as is.

But this won't affect how `poll()` behaves as it will be an infinite timeout as long as it is negative

Reply:
Yes, these will produce different underlying values. I did this as the conversions from `i32/i16/i8` are able to be slightly more efficient while maintaining the same overall behavior. I can change them to be more like `i128/i64` if you want, I am not strongly opinionated here.

Reply:
I agree the current impl is more performant.

My worry is that users can convert a signed integer to a `PollTimeOut`, and they can also convert it back, then such an inconsistence would be discovered, which may confuses them a little bit:

```rs
let orig_value1: i128 = -2;
let orig_value2: i132 = -2;

let timeout1 = PollTimeOut::from(orig_value1);
let timeout2 = PollTimeOut::from(orig_value2);

let value1: i128 = timeout1.into();
let value2: i32 = timeout2.into();

assert_eq!(orig_value1, value1); // assert_eq!(-2, -1);
assert_eq!(orig_value2, value2); // assert_eq!(-2, -2);
```

Reply:
Would documentation on `PollTimeout` be sufficient to alleviate this concern?

Reply:
Yeap, a documentation on this behavior would be nice:)

Reply:
Actually, accepting negative values for `timeout` seems to be a Linuxism.  POSIX only requires that `-1` be interpreted as "forever", and FreeBSD will return `EINVAL` for other negative values.  Please make it impossible to contruct a `PollTimeout` with a negative value other than `-1`.

Reply:
It is quite a bit messier to encapsulate this logic as it requires custom `Error` types for many of the `TryFrom` implementations to cover this.

Reply:
> I do not think the POSIX specification specifies that negative values other than -1 should error. 

I think @asomers was not saying that specifying a negative value other than -1 should be considered an error in the POSIX standard, it is the FreeBSD implementation that treats that case as an error

The behavior of FreeBSD is documented in their [manual](https://man.freebsd.org/cgi/man.cgi?poll#end), and from this issue [cpython: select.poll.poll fails on BSDs with arbitrary negative timeouts](https://github.com/python/cpython/issues/75515), this seems to be a general behavior across BSD OSes

Reply:
Okay, I have re-worked the conversions to return an error when a non -1 negative is given.